Sources/NIOExtras/LengthFieldBasedFrameDecoder.swift

```diff
@@ -0,0 +1,122 @@
+///
+///     +-----+------+
+///     | BCD | FGHI |
+///     +-----+------+
+///
+/// 'A' and 'E' will be the headers and will not be passed forward.
+///
+
+public final class LengthFieldBasedFrameDecoder: ByteToMessageDecoder {
+
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    
+    public var cumulationBuffer: ByteBuffer?
+    
+    private let lengthFieldLength: Int
+    private let lengthFieldEndianness: Endianness
+    
+    /// Create `LengthFieldBasedFrameDecoder` with a given frame length.
+    ///
+    /// - parameters:
+    ///    - lengthFieldLength: The length of the field specifying the remaining length of the frame.
+    ///    - lengthFieldEndianness: The endianness of the field specifying the remaining length of the frame.
+    ///
+    public init(lengthFieldLength: Int, lengthFieldEndianness: Endianness = .little) {
+        
+        precondition(lengthFieldLength >= 0, "lengthField length must not be negative")
+        precondition(lengthFieldLength > 0, "lengthField length must not be zero")
+        precondition(lengthFieldLength <= 8, "lengthField length only handles up to 64bit (8 byte)")
+
+        self.lengthFieldLength = lengthFieldLength
+        self.lengthFieldEndianness = lengthFieldEndianness
+    }
+    
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+        
+        guard let lengthFieldSlice = buffer.readSlice(length: lengthFieldLength) else {
+            return .needMoreData
+        }
+
+        // convert the length field to an int specifying the length
+        guard let lengthFieldValue = frameLength(for: lengthFieldSlice,
+                                                 length: lengthFieldLength,
+                                                 endianness: lengthFieldEndianness) else {
+            return .needMoreData
+        }
+
+        guard let contentsFieldSlice = buffer.readSlice(length: lengthFieldValue) else {
+            return .needMoreData
+        }
+
+        ctx.fireChannelRead(self.wrapInboundOut(contentsFieldSlice))
+
+        return .continue
+    }
+    
+    public func handlerRemoved(ctx: ChannelHandlerContext) {
+        if let buffer = cumulationBuffer, buffer.readableBytes > 0 {
+            ctx.fireErrorCaught(NIOExtrasErrors.LeftOverBytesError(leftOverBytes: buffer))
+        }
+    }
+
+    ///
+    /// Decodes the specified region of the buffer into an unadjusted frame length. The default implementation is
+    /// capable of decoding the specified region into an unsigned 8/16/32/64 bit integer.
+    /// - parameters:
+    ///    - buffer: The buffer containing the integer frame length
+    ///    - length: The length of the integer contained within the buffer.
+    ///    - endianness: The endianness of the integer contained within the buffer.
+    ///
+    private func frameLength(for buffer: ByteBuffer, length: Int, endianness: Endianness) -> Int? {
```

Comment:
`endianness` doesn't need to be a parameter to this function, it's stored on the class.

Reply:
will remove